ChatWithTree/ChatWithTreeBot.py

```diff
@@ -0,0 +1,664 @@
+            # yield from returns all yields from the calling func
+            yield from self.get_chatbot_response(message)
+        else:
+            yield (YieldType.FINAL, "Error: ensure to set GRAMPS_AI_MODEL_NAME and GRAMPS_AI_MODEL_URL environment variables.")
+      
+    
+
+   # @_throttle.rate_limited(_limiter)
+    def _llm_complete(
+        self,
+        all_messages: List[Dict[str, str]],
+        tool_definitions: Optional[List[Dict[str, str]]],
+        seed: int,
+    ) -> Any:
+        response = litellm.completion(
+            model=GRAMPS_AI_MODEL_NAME,  # self.model,
+            messages=all_messages,
+            seed=seed,
+            tools=tool_definitions,
+            tool_choice="auto" if tool_definitions is not None else None,
+        )
+
+        # logger.debug("\033[92mResponse from AI Model:\033[0m")
+        # Convert response to a dictionary if possible
+        response_dict = response.to_dict() if hasattr(response, 'to_dict') else str(response)
+        # logger.debug(json.dumps(response_dict, indent=2))
+        return response
+
+    def get_chatbot_response(
+        self,
+        user_input: str,
+        seed: int = 42,
+    ) -> Iterator[Tuple[YieldType, str]]:
+        self.messages.append({"role": "user", "content": user_input})
+        yield from self._llm_loop(seed)
+
+    def execute_tool(self, tool_call):
+        # logger.debug(f"Executing tool call: {tool_call['function']['name']}")
+        tool_name = tool_call["function"]["name"]
+        arguments = json.loads(tool_call["function"]["arguments"])
+        sys.stdout.flush()
+        tool_func = self.tool_map.get(tool_name)
+        try:
+            if tool_func is not None:
+                sig = inspect.signature(tool_func)
+                if len(sig.parameters) == 0:
+                    # Ignore any arguments, call with none
+                    tool_result = tool_func()
+                else:
+                    tool_result = tool_func(**arguments)
+
+            else:
+                tool_result = f"Unknown tool: {tool_name}"
+
+            content_for_llm = ""
+            if isinstance(tool_result, (dict, list)):
+                content_for_llm = json.dumps(tool_result)
+            else:
+                content_for_llm = str(tool_result)
+
+            #logger.debug("\033[93mTool call result:\033[0m")
+            #logger.debug(content_for_llm)
+
+        except Exception as exc:
+            #logger.debug(exc)
+            content_for_llm = f"Error in calling tool `{tool_name}`: {exc}"  # Include exception for LLM clarity
+
+        self.messages.append(
+            {
+                "role": "tool",
+                "tool_call_id": tool_call["id"],
+                "content": content_for_llm,
+            }
+        )
+
+    def _llm_loop(self, seed: int) -> Iterator[Tuple[YieldType, str]]:
+        # Tool-calling loop
+        final_response = "I was unable to find the desired information."
+        limit_loop = 6
```

Comment:
Not sure that will be enough. Maybe make this a parameter, or config setting?

Reply:
/setlimit is now implemented